blogposts/2021/postgres-version-update.md

```diff
@@ -0,0 +1,139 @@
+
+**If you are maintaining an external database**: If your Postgres version is older than Postgres 12, you will need to update your database instance prior to upgrading from Sourcegraph 3.26 to 3.27.
+
+## Why does Sourcegraph 3.27 require Postgres 12?
+
+Postgres 9.6 was released on 2016/09/29 and since then there have been a [flood of changes](https://www.postgresql.org/about/featurematrix/). Among these changes is a feature which Sourcegraph requires moving forward: Postgres 9.6 contains a SQL standard compatibility exclusion, as noted in the documentation for [creating triggers](https://www.postgresql.org/docs/9.6/sql-createtrigger.html#SQL-CREATETRIGGER-COMPATIBILITY):
+
+> PostgreSQL does not allow the old and new tables to be referenced in statement-level triggers, i.e., the tables that contain all the old and/or new rows, which are referred to by the OLD TABLE and NEW TABLE clauses in the SQL standard.
+
+Postgres 12, however, supports new and old table references in statement-level triggers.
+
+The rest of this post explains the technical motivation for why we needed this capability.
+
+### What does this mean?
+
+In SQL, there are two ways a trigger can be executed:
+
+1. Before or after a _row_ is inserted into/updated within/deleted from a target table
+1. Before or after an insert, update, or delete _statement_ over a target table
+
+In the former case, the trigger has access to the `OLD` and `NEW` row values. This allows the trigger to perform actions conditionally as well as extract the previous and updated values in the case the trigger is used to keep denormalized data consistent.
+
+In the latter case, the trigger does not have access to the previous or the new data. Unfortunately, this means that statement-level triggers have no access to previous data, and must re-query the new data from the target table to make use of it.
+
+### Why do we need this particular feature?
+
+Sourcegraph performs database migrations on application startup. We try our best to keep these migrations fast, as a slow migration will block additional container instances from starting and accepting traffic.
+
+Some classes of migrations are very difficult to do efficiently:
+
+- Index creation over large tables requires a full table lock
+- Concurrent index creation cannot be applied within a transaction
+- Concurrent index creation takes much longer, can be blocked for an arbitrary amount of time due to live transactions, and can fail requiring an operator to rebuild the index (cherry on top: reindex is not concurrent)
+- Data migrations over large tables can take an arbitrary amount of time
+
+Sourcegraph 3.25 introduced a mechanism for performing out-of-band migrations. Instead of being applied synchronously on application startup, out-of-band migrations are applied in the application background over a longer period of time.
```

Comment:
```suggestion
[Sourcegraph 3.25](/blog/release/3.25/) introduced a mechanism for performing out-of-band migrations. Instead of being applied synchronously on application startup, out-of-band migrations are applied in the application background over a longer period of time.
```